tests/core/test_datamodules.py

```diff
@@ -533,6 +535,81 @@ def test_simple_hyperparameters_saving():
     assert data.hparams == AttributeDict({"arg0": 10, "arg1": "foo", "kwarg0": "bar"})
 
 
+def test_define_as_dataclass():
+    """
+    Test datamodule for initialization as Python dataclass.
+    """
+
+    class RandomDataset(torch.utils.data.Dataset):
+        def __init__(self, size, length):
+            self.len = length
+            self.data = torch.randn(length, size)
+
+        def __getitem__(self, index):
+            return self.data[index]
+
+        def __len__(self):
+            return self.len
+
+    @dataclass(init=True, repr=True, eq=True, order=True, unsafe_hash=True, frozen=False)
+    class BoringDataModule1(LightningDataModule):
+        def __init__(self, batch_size):
+            super().__init__()
+            self.batch_size = batch_size
+
+        def train_dataloader(self):
+            return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=self.batch_size)
+
+    # assets for the different dunder methods added by dataclass, when __init__ is impelmented, i.e.
+    # __repr__, __eq__, __lt__, __le__, etc.
+    assert BoringDataModule1(batch_size=32)
+    assert hasattr(BoringDataModule1, "__repr__")
+    assert BoringDataModule1(batch_size=32) == BoringDataModule1(batch_size=32)
+
+    @dataclass(init=True, repr=True, eq=True, order=True, unsafe_hash=True, frozen=False)
+    class BoringDataModule2(LightningDataModule):
+
+        batch_size: int = 2
+
+        def train_dataloader(self):
+            return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=self.batch_size)
+
+    # assets for the different dunder methods added by dataclass, when __post_init__ is impelmented, i.e.
+    # __init__, __repr__, __eq__, __lt__, __le__, etc.
+    assert BoringDataModule2(batch_size=32)
+    assert hasattr(BoringDataModule2, "__repr__")
+    assert BoringDataModule2(batch_size=32) == BoringDataModule2(batch_size=32)
+
+    class BoringModel(LightningModule):
+        def __init__(self):
+            super().__init__()
+            self.layer = torch.nn.Linear(32, 2)
+
+        def forward(self, x):
+            return self.layer(x)
+
+        def training_step(self, batch, batch_idx):
+            loss = self(batch).sum()
+            self.log("train_loss", loss)
+            return {"loss": loss}
+
+        def configure_optimizers(self):
+            return torch.optim.SGD(self.layer.parameters(), lr=0.1)
+
+    trainer = Trainer(
+        default_root_dir=os.getcwd(),
+        limit_train_batches=1,
+        limit_val_batches=1,
+        num_sanity_val_steps=0,
+        max_epochs=1,
+        weights_summary=None,
+    )
+    trainer.fit(BoringModel(), datamodule=BoringDataModule2(batch_size=32))
```

Comment:
let's not run a training here. the above assertions for the datamodule are enough to determine if it's working well as a dataclass

Reply:
Done!